File: src/1396-Design-Underground-System/1396.js

var UndergroundSystem = function() {
    this.check_in = new Map();
    this.check_out = new Map();
};

/** 
 * @param {number} id 
 * @param {string} stationName 
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkIn = function(id, stationName, t) {
    this.check_in.set(id, [stationName, t]);
};

/** 
 * @param {number} id 
 * @param {string} stationName 
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkOut = function(id, stationName, t) {
    let [startName, timeIn] = this.check_in.get(id);
    let name = startName + ">" + stationName;
    if (!this.check_out.has(name)) this.check_out.set(name, [0, 0]);
    let [avg, cnt] = this.check_out.get(name);
    this.check_out.set(name, [avg + t - timeIn, cnt + 1]);
};

/** 
 * @param {string} startStation 
 * @param {string} endStation
 * @return {number}
 */
UndergroundSystem.prototype.getAverageTime = function(startStation, endStation) {
    let name = startStation + ">" + endStation;
    return this.check_out.get(name)[0] / this.check_out.get(name)[1];
};

File: src/1396-Design-Underground-System/1396.py
class UndergroundSystem:

    def __init__(self):
        self.check_in = collections.defaultdict(list)
        self.check_out = collections.defaultdict(list)


    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.check_in[id] = [stationName, t]


    def checkOut(self, id: int, stationName: str, t: int) -> None:
        startName, timeIn = self.check_in[id]
        name = startName + ">" + stationName
        if not self.check_out[name]: 
            self.check_out[name] = [0, 0]
        self.check_out[name][0] += t - timeIn
        self.check_out[name][1] += 1


    def getAverageTime(self, startStation: str, endStation: str) -> float:
        name = startStation + ">" + endStation
        return self.check_out[name][0] / self.check_out[name][1]